### ISSUE TITLE
`Z3_eval_smtlib2_string` overhead: can I do anything about it ?

### ISSUE THREAD

**Issue by qaristote** [Z3Prover/z3#6408]
[Here](https://gist.github.com/qaristote/1dc2ee7033c44a2dcd14713b47f80c54)'s an SMTLib2 file I want to evaluate with Z3 (I'm using version `4.8.15`). Using [`hyperfine`](https://github.com/sharkdp/hyperfine) (with `--runs 250` and `--warmup 5`) I benchmarked different ways of doing so:

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `z3 $input >/dev/null` | 181.8 ± 5.9 | 171.7 | 208.9 | 1.00 |
| `./bindings-oneshot.c.bin $input >/dev/null` | 189.2 ± 19.8 | 183.7 | 499.5 | 1.04 ± 0.11 |
| `./bindings.c.bin $input >/dev/null` | 253.3 ± 5.9 | 248.3 | 317.9 | 1.39 ± 0.06 |

[`bindings.c.bin`](https://gist.github.com/qaristote/6910cea4473dd7bc38b0eecd4cd4ae56) uses the `Z3_eval_smtlib2_string` function from the API to evaluate the input file line by line, while [`bindings-oneshot.c.bin`](https://gist.github.com/qaristote/2bb18ab65a3d7974ca9b19fd01e4f41a) does the same thing except the whole file is passed to Z3 in one go.

As you can see, there is a significant overhead in repeatedly running `Z3_eval_smtlib2_parse` on each line instead of running it once on the whole file. Is there a reason for that ? Is there anything I can do to have the line-by-line version be faster ?

**Comment by NikolajBjorner** [comment]
Mainly use the I/O pipe for parsing SMT commands.

The API function Z3_eval_smtlib2_string is not tuned for line-by-line: it re-initializes a parser every time it is invoked.
This is going to be significant overhead (or you can run a profiler to get source information). 
Tuning this code would require exposing a chunk that recycles the parser between calls.


Background.
This line is invoked from the API function:

https://github.com/Z3Prover/z3/blob/1fc77c8c004521867f1a78368f3c956c3ecfdbb8/src/parsers/smt2/smt2parser.cpp#L3209

It initializes the parser class.

Exposing the smt2::parser class (as a pointer) and re-invoking the parse operator is perhaps possible but requires dev/test. Specifically, it relies on the EOF token of the scanner to be reset should it be re-invoked.

**Comment by qaristote** [comment]
> Mainly use the I/O pipe for parsing SMT commands.

Just to make sure, are you saying I should be launching Z3 as an external process and sending commands by writing on its IO pipes ?

Anyways, thanks for the explanation !

**Comment by NikolajBjorner** [comment]
Right, the main uses of the SMTLIB2 text interface in other tools (such as Boogie, F*) is by spawning a process.
You could also speed up your processing by _not_ parsing single lines but batches of lines. As long as lines are declare-fun, just continue. Changing the smt2parser to allow repeated calls is possible, but I don't think it makes sense to take on liability of changing this code on my side. You can of course contribute by making this possible. A simple way to change it is by making the m_scanner field in the parser class a pointer (scoped_ptr<smt2::scanner> m_scanner), and adding a reset method where you pass in a new std::istream&.
